lib/fs.js

```diff
@@ -2479,3 +2479,452 @@ Object.defineProperty(fs, 'SyncWriteStream', {
   set: internalUtil.deprecate((val) => { SyncWriteStream = val; },
                               'fs.SyncWriteStream is deprecated.', 'DEP0061')
 });
+
+
+// The Promises API
+
+async function* readEverything(fd) {
+  let eof = false;
+  let position = 0;
+  const buffer = Buffer.allocUnsafe(kReadFileBufferLength);
+  while (!eof) {
+    const { length } =
+      await fs.promises.read(fd, buffer, 0, buffer.length, position);
+    if (length < buffer.length) {
+      eof = true;
+      return buffer.slice(0, length);  // Ended
+    } else {
+      position += length;
+      yield buffer.slice(0, length);
+    }
+  }
+}
+
+
+async function writeEverything(fd, buffer, offset, length, position) {
+  const { written } =
+    await fs.promises.write(fd, buffer, offset, length, position);
+  if (written !== length) {
+    offset += written;
+    length -= written;
+    if (position !== null)
+      position += written;
+    return writeEverything(fd, buffer, offset, length, position);
+  }
+}
+
+fs.promises = {
+  async access(path, mode = fs.F_OK) {
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+
+    if (typeof path !== 'string' && !(path instanceof Buffer)) {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                                 ['string', 'Buffer', 'URL']);
+    }
+
+    mode = mode | 0;
+    const req = new FSReqPromise();
+    binding.access(pathModule.toNamespacedPath(path), mode, req);
+    return req.promise;
+  },
+
+  async appendFile(path, data, options) {
+    callback = maybeCallback(callback || options);
+    options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'a' });
+    options = copyObject(options);
+    if (!options.flag || isFd(path))
+      options.flag = 'a';
+    return fs.promises.writeFile(path, data, options);
+  },
+
+  async chown(path, uid, gid) {
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+    const req = new FSReqPromise();
+    binding.chown(pathModule.toNamespacedPath(path), uid, gid, req);
+    return req.promise;
+  },
+
+  async chmod(path, mode) {
+    handleError((path = getPathFromURL(path)));
+    nullCheck(path);
+    const req = new FSReqPromise();
+    binding.chmod(pathModule.toNamespacedPath(path),
+                  modeNum(mode),
+                  req);
+    return req.promise;
+  },
+
+  async copyFile(src, dest, flags = 0) {
+    handleError((src = getPathFromURL(src)));
+    handleError((dest = getPathFromURL(dest)));
+    nullCheck(src);
+    nullCheck(dest);
+    src = pathModule.toNamespacedPath(src);
+    dest = pathModule.toNamespacedPath(dest);
+    flags = flags | 0;
+    const req = new FSReqPromise();
+    binding.copyFile(src, dest, flags, req);
+    return req.promise;
+  },
+
+  async fchmod(fd, mode) {
+    mode = modeNum(mode);
+    if (fd == null || typeof fd !== 'object' || !Number.isInteger(fd.fd))
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd', 'FD');
```

Comment:
Using [`ERR_INVALID_FD`](https://nodejs.org/dist/latest-v9.x/docs/api/errors.html#ERR_INVALID_FD) might be more suitable for the errors like this.